instrumentation/helpers-sql-obfuscation/lib/opentelemetry/instrumentation/helpers/sql_obfuscation.rb

```diff
@@ -0,0 +1,150 @@
+          double_quotes: /"(?:[^"]|"")*?(?:\\".*|"(?!"))/,
+          dollar_quotes: /(\$(?!\d)[^$]*?\$).*?(?:\1|$)/,
+          uuids: /\{?(?:[0-9a-fA-F]\-*){32}\}?/,
+          numeric_literals: /-?\b(?:[0-9]+\.)?[0-9]+([eE][+-]?[0-9]+)?\b/,
+          boolean_literals: /\b(?:true|false|null)\b/i,
+          hexadecimal_literals: /0x[0-9a-fA-F]+/,
+          comments: /(?:#|--).*?(?=\r|\n|$)/i,
+          multi_line_comments: %r{\/\*(?:[^\/]|\/[^*])*?(?:\*\/|\/\*.*)}
+        }.freeze
+
+        DIALECT_COMPONENTS = {
+          fallback: COMPONENTS_REGEX_MAP.keys,
+          mysql: %i[
+            single_quotes
+            double_quotes
+            numeric_literals
+            boolean_literals
+            hexadecimal_literals
+            comments
+            multi_line_comments
+          ],
+          postgres: %i[
+            single_quotes
+            dollar_quotes
+            uuids
+            numeric_literals
+            boolean_literals
+            comments
+            multi_line_comments
+          ]
+        }.freeze
+
+        PLACEHOLDER = '?'
+
+        # We use these to check whether the query contains any quote characters
+        # after obfuscation. If so, that's a good indication that the original
+        # query was malformed, and so our obfuscation can't reliably find
+        # literals. In such a case, we'll replace the entire query with a
+        # placeholder.
+        CLEANUP_REGEX = {
+          mysql: %r{'|"|\/\*|\*\/},
+          postgres: %r{'|\/\*|\*\/|\$(?!\?)}
+        }.freeze
+
+        # @api private
+        def self.generate_regex(dialect)
+          components = DIALECT_COMPONENTS[dialect]
+          Regexp.union(components.map { |component| COMPONENTS_REGEX_MAP[component] })
+        end
+
+        MYSQL_COMPONENTS_REGEX = generate_regex(:mysql)
+        POSTGRES_COMPONENTS_REGEX = generate_regex(:postgres)
+
+        # This is a SQL obfuscation utility intended for use in database adapter instrumentation.
+        #
+        # @param sql [String] The SQL to obfuscate.
+        # @param config [Hash] The user's configuration for the database adapter.
+        #   Required keys:
+        #     * `:db_statement` => A symbol with the values of either `:omit`, `:include`, or `:obfuscate`
+        #
+        #     * `:obfuscation_limit` => An integer representing the maximum length of a sql statement
+        #
+        #   Please include these keys in your instrumentation's configuration.
+        # @param adapter [Symbol] the type of database adapter calling the method. `:mysql` and `:postgres` are supported.
+        # @return [String] If successful, the SQL query string where the values are replaced with "?". If unsuccessful,
+        #  a string with a failure message.
+        #
+        # @api public
+        def self.obfuscate_sql(sql, config, adapter)
+          return sql unless config[:db_statement] == :obfuscate
+
+          regex = const_get("#{adapter.upcase}_COMPONENTS_REGEX")
```

Comment:
seems like we've sacrificed some performance on the hot path for the sake of future extensibility, if i'm understanding this code correctly? Haven't benchmarked any of this ofc and it's possible i'm wrong bc i ruby bad, but at a glance we seem to be passing in a symbol, then massaging it to an upcased string, whereas a hard coded case statement could avoid this. Do you see a drawback to that approach?

Reply:
I need to run perf tests on this too. IIRC, case statements are very slow so that's why I didn't go straight to that approach. I'll run some tests and add them to this thread.

Reply:
sure, it's possible a case statement has poor performance here, but it just seems like calling `.upcase` on a Symbol (https://ruby-doc.org/core-2.6.1/Symbol.html#method-i-upcase) is avoidable here, which under the hood is also calling `.to_s` on the Symbol, which would add a String instance allocation on every invocation iiuc?

Reply:
Your senses are correct! A case statement would be much more performant in this scenario and yields considerably fewer object allocations. 

Here's a gist with the test I ran: https://gist.github.com/kaylareopelle/ad910fd920401587fa0c4c79bbf7c98f

Reply:
Fixed in: 8096551319f684cf850901dbc12100dee43c8694